apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/runnable/GenerateBlocksThread.java

```diff
@@ -0,0 +1,158 @@
+                                    List<Block> poppedOffBlock = lookupBlockchainProcessor().popOffToCommonBlock(previousBlock);
+                                    for (Block block : poppedOffBlock) {
+                                        transactionProcessor.processLater(block.getOrLoadTransactions());
+                                    }
+                                    lastBlock = previousBlock;
+                                    lastBlockId = previousBlock.getId();
+                                    break;
+                                }
+                            }
+                        }
+                        List<GeneratorMemoryEntity> forgers = new ArrayList<>();
+                        for (GeneratorMemoryEntity generator : generatorsMap.values()) {
+                            generatorService.setLastBlock(lastBlock, generator);
+                            if (generator.getEffectiveBalance().signum() > 0) {
+                                forgers.add(generator);
+                            }
+                        }
+                        Collections.sort(forgers);
+                        sortedForgers = Collections.unmodifiableList(forgers);
+                        logged = false;
+                    }
+                    if (!logged) {
+                        for (GeneratorMemoryEntity generator : sortedForgers) {
+                            if (generator.getHitTime() - generationLimit > 60) {
+                                break;
+                            }
+                            log.debug(generator.toString());
+                            logged = true;
+                        }
+                    }
+                    for (GeneratorMemoryEntity generator : sortedForgers) {
+                        if (suspendForging) {
+                            break;
+                        }
+                        if (generator.getHitTime() > generationLimit
+                            || generatorService.forge(lastBlock, generationLimit, generator)) {
+                            return;
+                        }
+                    }
+                } finally {
+                    globalSync.updateUnlock();
+                }
+            } catch (Exception e) {
+                log.info("Error in block generation thread", e);
+            }
+        } catch (Throwable t) {
+            log.error("CRITICAL ERROR. PLEASE REPORT TO THE DEVELOPERS.\n" + t.toString());
+            t.printStackTrace();
+            System.exit(1);
+        }
+    }
+
+    public List<GeneratorMemoryEntity> getSortedForgers() {
+        return sortedForgers;
+    }
+
+    public void resetSortedForgers() {
+        sortedForgers = null;
+    }
+
+    public void setDelayTime(int delayTime) {
+        this.delayTime = delayTime;
+    }
+
+    private BlockchainProcessor lookupBlockchainProcessor() {
```

Comment:
Better to avoid lookups in refactored code

Reply:
Class uses both components like:
private final TransactionProcessor transactionProcessor;
private BlockchainProcessor blockchainProcessor;

but that's not possible inject them both via Constructor, because CDI finds 'cyclic dependency' (between BlockchainProcessor and TransactionProcessor).

So lookup is used to break it.